### PR TITLE
[20739] Refactor IStatusQueryable and make monitor service interfaces private

### PR DESCRIPTION
**Pull request by Mario-DL**
## Description

This PR brings to `master` the work from #4640 commit a2f466490c036c9062bef725693289891f7935fa and 7a0c3287b00a9ca2939896e3cf22c67ed757b2f2. It refactors the `IStatusQueryable()` interface and make monitor service interfaces private (except from `IProxyQueryable.hpp`  as `PDP.h` is still public in `master` and implements it).

@Mergifyio backport 2.13.x

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- **N/A** Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- **N/A** Any new/modified methods have been properly documented using Doxygen. 
- [X] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    Related documentation PR: eProsima/Fast-DDS-docs#745
- [X] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/subscriber/SubscriberImpl.hpp [hunk 1]
```diff
@@ -29,7 +29,7 @@
 #include <fastdds/dds/core/status/StatusMask.hpp>
 #include <fastrtps/types/TypesBase.h>
 
-#include <fastdds/statistics/rtps/monitor_service/interfaces/IStatusQueryable.hpp>
+#include <statistics/rtps/monitor-service/interfaces/IStatusQueryable.hpp>
 
 #include <mutex>
 #include <map>
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/fastdds/subscriber/SubscriberImpl.hpp:35]: Please, apply header order

File: src/cpp/statistics/rtps/monitor-service/MonitorService.cpp [hunk 2]
```diff
@@ -287,62 +262,28 @@ bool MonitorService::write_status(
                 status_data.local_entity(to_statistics_type({local_participant_guid_.guidPrefix, entity_id}));
                 GUID_t local_entity_guid = {local_participant_guid_.guidPrefix, entity_id};
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/statistics/rtps/monitor-service/MonitorService.cpp:263]: I am getting a warning building locally:

```
warning: variable ‘local_entity_guid’ set but not used [-Wunused-but-set-variable]
  263 |                 GUID_t local_entity_guid = {local_participant_guid_.guidPrefix, entity_id};
```

The variable is used but not in every case. In order to remove the warning, please use the variable casting it to void.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
Two nitpicks

**Comment by Mario-DL** [comment]
@richiprosima please test this

File: src/cpp/fastdds/subscriber/SubscriberImpl.hpp [hunk 3]
```diff
@@ -23,16 +23,16 @@
...
 #include <fastdds/dds/subscriber/qos/DataReaderQos.hpp>
 #include <fastdds/dds/subscriber/qos/SubscriberQos.hpp>
-#include <fastdds/dds/core/status/StatusMask.hpp>
 #include <fastrtps/types/TypesBase.h>
 
 #include <statistics/rtps/monitor-service/interfaces/IStatusQueryable.hpp>
 
-#include <mutex>
 #include <map>
+#include <mutex>
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/fastdds/subscriber/SubscriberImpl.hpp]: The header order is not only alphabetical. First the STL headers, then external library headers and finally header of Fast DDS separated between public and private headers.

File: src/cpp/statistics/rtps/monitor-service/MonitorService.cpp [hunk 4]
```diff
@@ -262,73 +262,77 @@ bool MonitorService::write_status(
...
-                        assert(status_queryable_.get_monitoring_status(local_entity_guid, data));
+                        status_retrieved = status_queryable_.get_monitoring_status(local_entity_guid, data);
                         break;
                     }
                     //Not triggered for the moment
                     case INCONSISTENT_TOPIC:
                     {
                         EPROSIMA_LOG_ERROR(MONITOR_SERVICE, "Inconsistent topic status not supported yet");
-                        assert(false);
+                        (void)local_entity_guid;
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/statistics/rtps/monitor-service/MonitorService.cpp:294]: I suggest a C++ casting instead of a C one.

File: src/cpp/statistics/rtps/monitor-service/MonitorService.cpp [hunk 5]
```diff
@@ -262,73 +262,77 @@ bool MonitorService::write_status(
...
                         break;
                     }
                 }
 
-                status_data.status_kind((StatusKind)i);
-                status_data.value(data);
-                add_change(status_data, false);
-
+                assert(status_retrieved);
+                if (status_retrieved)
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/statistics/rtps/monitor-service/MonitorService.cpp:330]: I suggest removing the condition if it is being asserted just above.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

File: src/cpp/statistics/rtps/monitor-service/MonitorService.cpp (hunk 5 above)
Comment (Mario-DL) [reply src/cpp/statistics/rtps/monitor-service/MonitorService.cpp:330]: If depending on the compilation the `assert` is removed, we should then avoid `add_change` if status could not be correctly retrieved. I think we should keep both. I will add a log in case is `false` and it could not be retrieved

**Comment by Mario-DL** [comment]
@richiprosima please test this

File: src/cpp/statistics/rtps/monitor-service/MonitorService.cpp (hunk 5 above)
Comment (JLBuenoLopez-eProsima) [reply src/cpp/statistics/rtps/monitor-service/MonitorService.cpp:330]: I have talked with Edu and he agrees with me that asserting and using a conditional checking that condition is incompatible. Either we are completely sure that a specific situation should never happen and therefore an assertion is done, or if there is some small possibility that a corner case occurs, then the condition should be kept.

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Review by EduPonz — CHANGES_REQUESTED** [review]
@Mario-DL please address documentation build failure

**Review by EduPonz — APPROVED** [review]
LGTM with green CI

**Comment by EduPonz** [comment]
I'm seeing the same Github Windows errors in nightlies and other PR and seem to have something to do with the runners infra, so I'm merging this one.

**Comment by EduPonz** [comment]
@Mergifyio backport 2.13.x
